Added support to different diagrams per XML file

snakebyte85 opened the pull request:
Different diagrams (processes or collaborations) can be saved inside a single XML file, following BPMN 2.0 standard, but bpmn-js renders only the first diagram.

This commit allows a developer to select a specific diagram to be rendered, using the ID of the process or collaboration in this simply way:

`bpmnViewer.importXML(bpmnXML, done, processId);`

The first diagram (like before) will be rendered if processId is null

bpmn-js internally already supported the selection of a specific diagram to render, but the interfaces did not allow this

This is useful for #87

philippfromme commented:
Thanks for filing this pull request. Being able to specify which diagram you want to import would definitely be an improvement. We'll look into this. In the meantime, a few thoughts:

* the specified ID should always reference a `bpmndi:BPMNDiagram`
* `importXML` could support specifying a model instance when opening a diagram so you can switch between different diagrams in an XML file without reimporting the XML every single time - alternatively we could use `importDefinitions`

snakebyte85 commented:
- select diagram only for diagramId
- Created method to switch diagram for an already imported XML in this way
`switchDiagram("<diagramId>")`

Let me know if something else is needed

nikku commented:
Thanks for your follow up @snakebyte85. Could you add some test cases that bring the code coverage to a decent amount?

snakebyte85 commented:
I added the needed test cases. I added two bpmns to be tested:

- One simple BPMN (with two diagrams) created by me
- Ones that comes from https://github.com/bpmn-miwg/bpmn-miwg-test-suite/blob/master/Reference/C.4.0.bpmn as official test case

Sorry if I didn't that before but I am not practical of the used test enviroment and used patterns.

Let me know if tests needed to be modified

barmac commented on lib/Viewer.js:
```diff
@@ -153,8 +154,9 @@ inherits(Viewer, Diagram);
+ * @param {String} ID of the diagram to render (if null or non-existent, the first will be rendered)
  */
-Viewer.prototype.importXML = function(xml, done) {
+Viewer.prototype.importXML = function(xml, done, diagramId) {
```
Can you please keep the current pattern of placing the callback at the end?
```suggestion
Viewer.prototype.importXML = function(xml, diagramId, done) {
```

snakebyte85 replied:
Are you sure is this needed?

The drawback to follow this pattern is that it would break all existing external code (e.g. camunda modeler) using this interface.

philippfromme replied:
No, you wouldn't, if `diagramId` is not present, `done` will be considered the second parameter.

barmac replied:
```
function check(first, optional, done) {
  if (isFunction(optional)) {
    done = optional;
    optional = null;
  }
  ...
}
```

barmac commented on test/spec/import/ImporterSpec.js:
```diff
@@ -40,12 +40,17 @@ describe('import - Importer', function() {
-      importBpmnDiagram(diagram, definitions, done);
+
+      var selectedDiagram = find(definitions.diagrams, function(element) {
+        return element.id == diagramId;
```
```suggestion
        return element.id === diagramId;
```

barmac commented on lib/Viewer.js:
```diff
@@ -185,16 +187,51 @@ Viewer.prototype.importXML = function(xml, done) {
+  var diagram = null;
+  if (diagramId) {
+    diagram = find(self._definitions.diagrams, function(element) {
+      return element.id == diagramId;
```
```suggestion
      return element.id === diagramId;
```

barmac commented on lib/Viewer.js:
```diff
@@ -185,16 +187,51 @@ Viewer.prototype.importXML = function(xml, done) {
+
+  self.importDefinitions(self._definitions, function(err, importWarnings) {
+
+    self._emit('import.done', { error: err, warnings: importWarnings });
```
Diagram switch is not an import so it should not fire the `import.done` event. Let's use the render events only for such a case.

barmac commented on lib/Viewer.js:
```diff
@@ -185,16 +187,51 @@ Viewer.prototype.importXML = function(xml, done) {
+/**
+ * Switch to another diagram of the same XML
+ *
+ * @param {String} ID of the diagram to render (if null or non-existent, the first will be rendered)
```
I am not sure if we should actually let to use a non-existent diagram ID and still notify the user a successful switch (or import). This should be a place to complete with an error as displaying the whole BPMN is not what user expects when they pass a diagram ID.

barmac commented on lib/Viewer.js:
```diff
@@ -185,16 +187,51 @@ Viewer.prototype.importXML = function(xml, done) {
+      diagram = find(definitions.diagrams, function(element) {
+        return element.id == diagramId;
+      });
+    }
```
This part is reoccurring in the code so it could be extracted to a separate function.

barmac commented:
Thank you for your further commits. I left a few comments that will need to be resolved before we merge the PR.

Can you please make sure your commits adhere to [our contributing rules](https://github.com/bpmn-io/bpmn-js/blob/master/.github/CONTRIBUTING.md#creating-a-pull-request)?

Can we also have a test case for a situation when the selected diagram is not present?

snakebyte85 commented:
All done

Let me know for something else

barmac commented:
I slightly changed the API so that instead of `#switchDiagram` we have `#open`. Also, callback will be invoked with real errors instead of objects with `message` property. Furthermore, I added documentation.

@nikku Can you please have a look at this?

nikku commented:
@snakebyte85 Thanks again for your contribution.

@barmac I discovered an issue with re-opens if elements are shown in multiple `BPMNDiagram` instances. Cf. https://github.com/bpmn-io/bpmn-js/commit/95c1895781fdd7ee4f9f903198c4e63e5590561f for a test case that fails. We'd need to sort out this issue before merging this feature. Could you have a look into it?

barmac commented:
Current state is on https://github.com/bpmn-io/bpmn-js/tree/select-diagram

I pushed a bugfix for BpmnTreeWalker when context wasn't passed. This revealed that the current implementation renders [more elements than enough](https://github.com/bpmn-io/bpmn-js/blob/select-diagram/test/spec/ViewerSpec.js#L827), as a result of [this](https://github.com/bpmn-io/bpmn-js/blob/master/lib/import/BpmnTreeWalker.js#L222).

There is also a problem of unnecessary warnings when an element is included in more than one DI.

barmac commented:
I rebased the branch on top of master. There are still two failing test cases.

1. We never deregister elements DI [once it's registered](https://github.com/bpmn-io/bpmn-js/blob/master/lib/import/BpmnTreeWalker.js#L108), which for some diagrams results in unnecessary rendering [here](https://github.com/bpmn-io/bpmn-js/blob/master/lib/import/BpmnTreeWalker.js#L222) during diagram switch. See [this test case](https://github.com/bpmn-io/bpmn-js/blob/select-diagram/test/spec/ViewerSpec.js#L882) for details.

2. At the moment, we send a warning if an element's DI prop has already been registered. This is noticeable during a render of second diagram which uses the same elements, as in [this test case](https://github.com/bpmn-io/bpmn-js/blob/select-diagram/test/spec/ViewerSpec.js#L848). I think if we fix the first issue, the second one might be resolved, too.

Please feel free to update your branch with `bpmn-js/select-diagram` and add solution for the abovementioned issues :).
